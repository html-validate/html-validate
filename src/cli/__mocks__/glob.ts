import minimatch from "minimatch";

const glob: any = jest.requireActual("glob");

interface Options {
	cwd?: string;
}

let mockFiles: string[] = null;

function setMockFiles(files: string[]): void {
	mockFiles = files;
}

function resetMock(): void {
	mockFiles = null;
}

const originalSync = glob.sync;
function syncMock(pattern: string, options: Options = {}): string[] {
	if (!mockFiles) {
		return originalSync(pattern, options);
	}

	/* slice the cwd away since it is prepended automatically and makes it harder
	 * to test with */
	const cwd = options.cwd || "";
	const dir = cwd.replace(process.cwd(), "").replace(/^\/(.*)/, "$1/");

	let src = mockFiles;
	if (dir) {
		src = src
			.filter(cur => cur.startsWith(dir))
			.map(cur => cur.slice(dir.length));
	}

	return src.filter(cur => minimatch(cur, pattern));
}

glob.setMockFiles = setMockFiles;
glob.resetMock = resetMock;
glob.sync = syncMock;

module.exports = glob;
