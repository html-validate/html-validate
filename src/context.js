module.exports = Context;

var Reporter = require('./reporter');

function Context(str, globalListeners){
	this.report = new Reporter();
	this.state = 0;
	this.string = str;
	this.stack = [];
	this.listeners = {};
	this.globalListeners = globalListeners;
}

Context.prototype.match = function(x){
	return this.string.match(x);
};

Context.prototype.consume = function(n, state){
	if ( typeof(n) !== 'number' ){
		n = n[0].length; /* regex match */
	}
	this.string = this.string.substr(n);

	if ( typeof(state) !== 'undefined' ){
		this.state = state;
	}
};

Context.prototype.push = function(node){
	this.stack.push(node);
};

Context.prototype.pop = function(){
	return this.stack.pop();
};

Context.prototype.top = function(n){
	n = n || 0;
	return this.stack[this.stack.length - (1 + n)];
};

Context.prototype.addRule = function(rule, options){
	rule.init(this, options);
};

Context.prototype.addListener = function(event, rule, callback){
	this.listeners[event] = this.listeners[event] || [];
	this.listeners[event].push({
		rule: rule,
		callback: callback,
	});
};

Context.prototype.trigger = function(eventname, data){
	var report = this.report;
	var event = Object.assign({event: eventname}, data);

	/* execute rule listeners */
	var listeners = this.listeners[eventname] || [];
	listeners.forEach(function(listener){
		var rule = listener.rule;
		listener.callback.call(rule, data, function(node, message){
			report.add(node, rule, message);
		});
	});

	/* execute any global listener */
	var globalListeners = [].concat(this.globalListeners[eventname] || [], this.globalListeners['*'] || []);
	globalListeners.forEach(function(listener){
		listener(event);
	});
};

Context.prototype.saveReport = function(dst){
	this.report.save(dst);
};
