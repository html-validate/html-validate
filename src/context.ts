'use strict';

export interface Source {
	data: string;
	filename: string;
}

export interface LocationData {
	filename: string;
	line: number;
	column: number;
}

export class Context {
	state: number;
	string: string;
	filename: string;
	line: number;
	column: number;

	constructor(source: Source){
		this.state = undefined;
		this.string = source.data;
		this.filename = source.filename;
		this.line = 1;
		this.column = 1;
	}

	consume(n: number|Array<string>, state?: number){
		/* if "n" is an regex match the first value is the full matched
		 * string so consume that many characters. */
		if ( typeof(n) !== 'number' ){
			n = n[0].length; /* regex match */
		}

		/* poor mans line counter :( */
		let consumed = this.string.slice(0, n);
		let offset;
		while ( (offset=consumed.indexOf('\n')) >= 0 ){
			this.line++;
			this.column = 1;
			consumed = consumed.substr(offset + 1);
		}
		this.column += consumed.length;

		/* remove N chars */
		this.string = this.string.substr(n);

		/* change state */
		if ( typeof(state) !== 'undefined' ){
			this.state = state;
		}
	}

	getLocationData(): LocationData {
		return {
			filename: this.filename,
			line: this.line,
			column: this.column,
		};
	}

	getLocationString(): string {
		return `${this.filename}:${this.line}:${this.column}`;
	}
}

export default Context;
