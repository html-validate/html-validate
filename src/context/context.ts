import { type Location } from "./location";
import { type Source } from "./source";
import { State } from "./state";

export enum ContentModel {
	TEXT = 1,
	SCRIPT,
	STYLE,
}

/**
 * @internal
 */
export class Context {
	public contentModel: ContentModel;
	public state: State;
	public string: string;
	private filename: string;
	private offset: number;
	private line: number;
	private column: number;

	public constructor(source: Source) {
		this.state = State.INITIAL;
		this.string = source.data;
		this.filename = source.filename;
		this.offset = source.offset;
		this.line = source.line;
		this.column = source.column;
		this.contentModel = ContentModel.TEXT;
	}

	public getTruncatedLine(n: number = 13): string {
		return JSON.stringify(this.string.length > n ? `${this.string.slice(0, 10)}...` : this.string);
	}

	public consume(n: number | string[], state: State): void {
		/* if "n" is an regex match the first value is the full matched
		 * string so consume that many characters. */
		if (typeof n !== "number") {
			n = n[0].length; /* regex match */
		}

		/* poor mans line counter :( */
		let consumed = this.string.slice(0, n);
		let offset;
		while ((offset = consumed.indexOf("\n")) >= 0) {
			this.line++;
			this.column = 1;
			consumed = consumed.substr(offset + 1);
		}
		this.column += consumed.length;
		this.offset += n;

		/* remove N chars */
		this.string = this.string.substr(n);

		/* change state */
		this.state = state;
	}

	public getLocation(size: number): Location {
		return {
			filename: this.filename,
			offset: this.offset,
			line: this.line,
			column: this.column,
			size,
		};
	}
}

export default Context;
