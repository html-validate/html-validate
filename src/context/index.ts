export { Source, ProcessElementContext } from "./source";
export { Location, sliceLocation } from "./location";
export { Context, ContentModel } from "./context";
