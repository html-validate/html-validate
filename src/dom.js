'use strict';

const DOMNode = require('./domnode').default;

class DOM {
	constructor(){
		this.root = DOMNode.rootNode();
		this.active = this.root;
	}

	pushActive(node){
		this.active = node;
	}

	popActive(){
		this.active = this.active.parent;
	}

	getActive(){
		return this.active;
	}
}

module.exports = DOM;
