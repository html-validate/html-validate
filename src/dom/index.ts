export { Attribute } from "./attribute";
export { HtmlElement, NodeClosed } from "./htmlelement";
export { DOMNode } from "./domnode";
export { DOMTokenList } from "./domtokenlist";
export { DOMTree } from "./domtree";
export { DynamicValue } from "./dynamic-value";
export { NodeType } from "./nodetype";
export { TextNode } from "./text";
