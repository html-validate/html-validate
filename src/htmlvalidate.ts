import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
