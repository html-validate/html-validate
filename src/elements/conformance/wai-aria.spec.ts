/**
 * Conformance tests for WAI-ARIA specification
 * https://www.w3.org/TR/wai-aria
 */

import path from "node:path";
import { globSync } from "glob";
import "../../jest";
import { ariaNaming } from "../../rules/helper";
import { Source } from "../../context";
import { HtmlElement } from "../../dom";
import { HtmlValidate } from "../../htmlvalidate";
import waiAria from "./wai-aria";

const htmlvalidate = new HtmlValidate({
	root: true,
	extends: ["html-validate:recommended"],
	elements: ["html5"],
});

async function getElement(role: string): Promise<HtmlElement | null> {
	const source: Source = {
		data: `<div role="${role}"></role>`,
		filename: "inline",
		line: 1,
		column: 1,
		offset: 0,
	};
	const parser = await htmlvalidate.getParserFor(source);
	const doc = parser.parseHtml(source.data);
	return doc.querySelector("div");
}

async function naming(role: string): Promise<string | null> {
	const element = await getElement(role);
	if (!element) {
		return null;
	}
	return ariaNaming(element);
}

describe("ยง5 The Roles Model", () => {
	describe("ยง5.2.8.5 Roles Supporting Name from Content", () => {
		it.each(waiAria)("$role", async ({ role, naming: expectedNaming }) => {
			expect.assertions(1);
			const result = await naming(role);
			expect(result).toBe(expectedNaming);
		});
	});
});

describe("validator test-files", () => {
	const mapping = {
		"abstract-roles-prohibited.html": {
			"no-abstract-role": "error",
		},
	} as const;

	const fixtureDir = "test-files/wai-aria";
	const files = globSync("*.html", { cwd: fixtureDir }) as Array<keyof typeof mapping>;

	it.each(files)("%s", async (filename) => {
		expect.assertions(1);
		const filePath = path.join(fixtureDir, filename);
		const htmlvalidate = new HtmlValidate({
			root: true,
			rules: mapping[filename],
		});
		const report = await htmlvalidate.validateFile(filePath);
		expect(report).toMatchCodeframe();
	});
});
