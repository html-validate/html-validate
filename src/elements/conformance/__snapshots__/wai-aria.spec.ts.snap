// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator test-files abstract-roles-prohibited.html 1`] = `
"error: Role "command" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:10:16:
   8 |     ROLES: https://www.w3.org/TR/wai-aria-1.2/#abstract_roles
   9 |     -->
> 10 |     <div role="command" id="abstract-role-command">command</div>
     |                ^^^^^^^
  11 |     <div role="composite" id="abstract-role-composite">composite</div>
  12 |     <div role="input" id="abstract-role-input">input</div>
  13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
Selector: #abstract-role-command
error: Role "composite" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:11:16:
   9 |     -->
  10 |     <div role="command" id="abstract-role-command">command</div>
> 11 |     <div role="composite" id="abstract-role-composite">composite</div>
     |                ^^^^^^^^^
  12 |     <div role="input" id="abstract-role-input">input</div>
  13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
  14 |     <div role="range" id="abstract-role-range">range</div>
Selector: #abstract-role-composite
error: Role "input" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:12:16:
  10 |     <div role="command" id="abstract-role-command">command</div>
  11 |     <div role="composite" id="abstract-role-composite">composite</div>
> 12 |     <div role="input" id="abstract-role-input">input</div>
     |                ^^^^^
  13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
  14 |     <div role="range" id="abstract-role-range">range</div>
  15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
Selector: #abstract-role-input
error: Role "landmark" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:13:16:
  11 |     <div role="composite" id="abstract-role-composite">composite</div>
  12 |     <div role="input" id="abstract-role-input">input</div>
> 13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
     |                ^^^^^^^^
  14 |     <div role="range" id="abstract-role-range">range</div>
  15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
  16 |     <div role="section" id="abstract-role-section">section</div>
Selector: #abstract-role-landmark
error: Role "range" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:14:16:
  12 |     <div role="input" id="abstract-role-input">input</div>
  13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
> 14 |     <div role="range" id="abstract-role-range">range</div>
     |                ^^^^^
  15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
  16 |     <div role="section" id="abstract-role-section">section</div>
  17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
Selector: #abstract-role-range
error: Role "roletype" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:15:16:
  13 |     <div role="landmark" id="abstract-role-landmark">landmark</div>
  14 |     <div role="range" id="abstract-role-range">range</div>
> 15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
     |                ^^^^^^^^
  16 |     <div role="section" id="abstract-role-section">section</div>
  17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
  18 |     <div role="select" id="abstract-role-select">select</div>
Selector: #abstract-role-roletype
error: Role "section" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:16:16:
  14 |     <div role="range" id="abstract-role-range">range</div>
  15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
> 16 |     <div role="section" id="abstract-role-section">section</div>
     |                ^^^^^^^
  17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
  18 |     <div role="select" id="abstract-role-select">select</div>
  19 |     <div role="structure" id="abstract-role-structure">structure</div>
Selector: #abstract-role-section
error: Role "sectionhead" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:17:16:
  15 |     <div role="roletype" id="abstract-role-roletype">roletype</div>
  16 |     <div role="section" id="abstract-role-section">section</div>
> 17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
     |                ^^^^^^^^^^^
  18 |     <div role="select" id="abstract-role-select">select</div>
  19 |     <div role="structure" id="abstract-role-structure">structure</div>
  20 |     <div role="widget" id="abstract-role-widget">widget</div>
Selector: #abstract-role-sectionhead
error: Role "select" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:18:16:
  16 |     <div role="section" id="abstract-role-section">section</div>
  17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
> 18 |     <div role="select" id="abstract-role-select">select</div>
     |                ^^^^^^
  19 |     <div role="structure" id="abstract-role-structure">structure</div>
  20 |     <div role="widget" id="abstract-role-widget">widget</div>
  21 |     <div role="window" id="abstract-role-window">window</div>
Selector: #abstract-role-select
error: Role "structure" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:19:16:
  17 |     <div role="sectionhead" id="abstract-role-sectionhead">sectionhead</div>
  18 |     <div role="select" id="abstract-role-select">select</div>
> 19 |     <div role="structure" id="abstract-role-structure">structure</div>
     |                ^^^^^^^^^
  20 |     <div role="widget" id="abstract-role-widget">widget</div>
  21 |     <div role="window" id="abstract-role-window">window</div>
  22 |   </body>
Selector: #abstract-role-structure
error: Role "widget" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:20:16:
  18 |     <div role="select" id="abstract-role-select">select</div>
  19 |     <div role="structure" id="abstract-role-structure">structure</div>
> 20 |     <div role="widget" id="abstract-role-widget">widget</div>
     |                ^^^^^^
  21 |     <div role="window" id="abstract-role-window">window</div>
  22 |   </body>
  23 | </html>
Selector: #abstract-role-widget
error: Role "window" is abstract and must not be used (no-abstract-role) at test-files/wai-aria/abstract-roles-prohibited.html:21:16:
  19 |     <div role="structure" id="abstract-role-structure">structure</div>
  20 |     <div role="widget" id="abstract-role-widget">widget</div>
> 21 |     <div role="window" id="abstract-role-window">window</div>
     |                ^^^^^^
  22 |   </body>
  23 | </html>
  24 |
Selector: #abstract-role-window"
`;
